From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: corymhall <43035978+corymhall@users.noreply.github.com>
Date: Wed, 29 May 2024 09:11:43 -0400
Subject: [PATCH] Revert "r/aws_db_proxy: Change `auth` from `TypeList` to
 `TypeSet` as order is not significant."

This reverts commit 2db643f461c058fb7a9e9940afef240016412050.

diff --git a/internal/service/rds/proxy.go b/internal/service/rds/proxy.go
index 93b0f2a4d6..8ba13b8354 100644
--- a/internal/service/rds/proxy.go
+++ b/internal/service/rds/proxy.go
@@ -19,7 +19,6 @@ import (
 	"github.com/hashicorp/terraform-provider-aws/internal/errs"
 	"github.com/hashicorp/terraform-provider-aws/internal/errs/sdkdiag"
 	"github.com/hashicorp/terraform-provider-aws/internal/flex"
-	"github.com/hashicorp/terraform-provider-aws/internal/sdkv2"
 	tfslices "github.com/hashicorp/terraform-provider-aws/internal/slices"
 	tftags "github.com/hashicorp/terraform-provider-aws/internal/tags"
 	"github.com/hashicorp/terraform-provider-aws/internal/tfresource"
@@ -53,7 +52,7 @@ func resourceProxy() *schema.Resource {
 				Computed: true,
 			},
 			"auth": {
-				Type:     schema.TypeSet,
+				Type:     schema.TypeList,
 				Required: true,
 				Elem: &schema.Resource{
 					Schema: map[string]*schema.Schema{
@@ -88,7 +87,6 @@ func resourceProxy() *schema.Resource {
 						},
 					},
 				},
-				Set: sdkv2.SimpleSchemaSetFunc("auth_scheme", names.AttrDescription, "iam_auth", "secret_arn", names.AttrUsername),
 			},
 			"debug_logging": {
 				Type:     schema.TypeBool,
@@ -149,7 +147,7 @@ func resourceProxyCreate(ctx context.Context, d *schema.ResourceData, meta inter
 
 	name := d.Get(names.AttrName).(string)
 	input := &rds.CreateDBProxyInput{
-		Auth:         expandUserAuthConfigs(d.Get("auth").(*schema.Set).List()),
+		Auth:         expandUserAuthConfigs(d.Get("auth").([]interface{})),
 		DBProxyName:  aws.String(name),
 		EngineFamily: types.EngineFamily(d.Get("engine_family").(string)),
 		RoleArn:      aws.String(d.Get(names.AttrRoleARN).(string)),
@@ -226,7 +224,7 @@ func resourceProxyUpdate(ctx context.Context, d *schema.ResourceData, meta inter
 	if d.HasChangesExcept(names.AttrTags, names.AttrTagsAll) {
 		oName, nName := d.GetChange(names.AttrName)
 		input := &rds.ModifyDBProxyInput{
-			Auth:           expandUserAuthConfigs(d.Get("auth").(*schema.Set).List()),
+			Auth:           expandUserAuthConfigs(d.Get("auth").([]interface{})),
 			DBProxyName:    aws.String(oName.(string)),
 			DebugLogging:   aws.Bool(d.Get("debug_logging").(bool)),
 			NewDBProxyName: aws.String(nName.(string)),
diff --git a/internal/service/rds/proxy_test.go b/internal/service/rds/proxy_test.go
index 77f58e1e4e..402b85fb6c 100644
--- a/internal/service/rds/proxy_test.go
+++ b/internal/service/rds/proxy_test.go
@@ -40,18 +40,16 @@ func TestAccRDSProxy_basic(t *testing.T) {
 		Steps: []resource.TestStep{
 			{
 				Config: testAccProxyConfig_basic(rName),
-				Check: resource.ComposeAggregateTestCheckFunc(
+				Check: resource.ComposeTestCheckFunc(
 					testAccCheckProxyExists(ctx, resourceName, &v),
 					resource.TestCheckResourceAttr(resourceName, names.AttrName, rName),
 					resource.TestCheckResourceAttr(resourceName, "engine_family", "MYSQL"),
 					acctest.MatchResourceAttrRegionalARN(resourceName, names.AttrARN, "rds", regexache.MustCompile(`db-proxy:.+`)),
-					resource.TestCheckResourceAttr(resourceName, "auth.#", acctest.Ct1),
-					resource.TestCheckTypeSetElemNestedAttrs(resourceName, "auth.*", map[string]string{
-						"auth_scheme":               "SECRETS",
-						"client_password_auth_type": "MYSQL_NATIVE_PASSWORD",
-						names.AttrDescription:       "test",
-						"iam_auth":                  "DISABLED",
-					}),
+					resource.TestCheckResourceAttr(resourceName, "auth.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "auth.0.auth_scheme", "SECRETS"),
+					resource.TestCheckResourceAttr(resourceName, "auth.0.description", "test"),
+					resource.TestCheckResourceAttr(resourceName, "auth.0.iam_auth", "DISABLED"),
+					resource.TestCheckResourceAttr(resourceName, "auth.0.client_password_auth_type", "MYSQL_NATIVE_PASSWORD"),
 					resource.TestCheckResourceAttr(resourceName, "debug_logging", acctest.CtFalse),
 					resource.TestMatchResourceAttr(resourceName, names.AttrEndpoint, regexache.MustCompile(`^[\w\-\.]+\.rds\.amazonaws\.com$`)),
 					resource.TestCheckResourceAttr(resourceName, "idle_client_timeout", "1800"),
@@ -332,10 +330,7 @@ func TestAccRDSProxy_authDescription(t *testing.T) {
 				Config: testAccProxyConfig_name(rName, rName),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckProxyExists(ctx, resourceName, &dbProxy),
-					resource.TestCheckResourceAttr(resourceName, "auth.#", acctest.Ct1),
-					resource.TestCheckTypeSetElemNestedAttrs(resourceName, "auth.*", map[string]string{
-						names.AttrDescription: "test",
-					}),
+					resource.TestCheckResourceAttr(resourceName, "auth.0.description", "test"),
 				),
 			},
 			{
@@ -347,10 +342,7 @@ func TestAccRDSProxy_authDescription(t *testing.T) {
 				Config: testAccProxyConfig_authDescription(rName, description),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckProxyExists(ctx, resourceName, &dbProxy),
-					resource.TestCheckResourceAttr(resourceName, "auth.#", acctest.Ct1),
-					resource.TestCheckTypeSetElemNestedAttrs(resourceName, "auth.*", map[string]string{
-						names.AttrDescription: description,
-					}),
+					resource.TestCheckResourceAttr(resourceName, "auth.0.description", description),
 				),
 			},
 		},
@@ -378,10 +370,7 @@ func TestAccRDSProxy_authIAMAuth(t *testing.T) {
 				Config: testAccProxyConfig_name(rName, rName),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckProxyExists(ctx, resourceName, &dbProxy),
-					resource.TestCheckResourceAttr(resourceName, "auth.#", acctest.Ct1),
-					resource.TestCheckTypeSetElemNestedAttrs(resourceName, "auth.*", map[string]string{
-						"iam_auth": "DISABLED",
-					}),
+					resource.TestCheckResourceAttr(resourceName, "auth.0.iam_auth", "DISABLED"),
 				),
 			},
 			{
@@ -393,10 +382,7 @@ func TestAccRDSProxy_authIAMAuth(t *testing.T) {
 				Config: testAccProxyConfig_authIAMAuth(rName, iamAuth),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckProxyExists(ctx, resourceName, &dbProxy),
-					resource.TestCheckResourceAttr(resourceName, "auth.#", acctest.Ct1),
-					resource.TestCheckTypeSetElemNestedAttrs(resourceName, "auth.*", map[string]string{
-						"iam_auth": iamAuth,
-					}),
+					resource.TestCheckResourceAttr(resourceName, "auth.0.iam_auth", iamAuth),
 				),
 			},
 		},
@@ -424,8 +410,7 @@ func TestAccRDSProxy_authSecretARN(t *testing.T) {
 				Config: testAccProxyConfig_name(rName, rName),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckProxyExists(ctx, resourceName, &dbProxy),
-					resource.TestCheckResourceAttr(resourceName, "auth.#", acctest.Ct1),
-					resource.TestCheckTypeSetElemAttrPair(resourceName, "auth.*.secret_arn", "aws_secretsmanager_secret.test", names.AttrARN),
+					resource.TestCheckResourceAttrPair(resourceName, "auth.0.secret_arn", "aws_secretsmanager_secret.test", names.AttrARN),
 				),
 			},
 			{
@@ -437,9 +422,7 @@ func TestAccRDSProxy_authSecretARN(t *testing.T) {
 				Config: testAccProxyConfig_authSecretARN(rName, nName),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckProxyExists(ctx, resourceName, &dbProxy),
-					resource.TestCheckResourceAttr(resourceName, "auth.#", acctest.Ct2),
-					resource.TestCheckTypeSetElemAttrPair(resourceName, "auth.*.secret_arn", "aws_secretsmanager_secret.test", names.AttrARN),
-					resource.TestCheckTypeSetElemAttrPair(resourceName, "auth.*.secret_arn", "aws_secretsmanager_secret.test2", names.AttrARN),
+					resource.TestCheckResourceAttrPair(resourceName, "auth.0.secret_arn", "aws_secretsmanager_secret.test2", names.AttrARN),
 				),
 			},
 		},
@@ -455,6 +438,8 @@ func TestAccRDSProxy_tags(t *testing.T) {
 	var dbProxy types.DBProxy
 	resourceName := "aws_db_proxy.test"
 	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
+	key := "foo"
+	value := "bar"
 
 	resource.ParallelTest(t, resource.TestCase{
 		PreCheck:                 func() { acctest.PreCheck(ctx, t); testAccDBProxyPreCheck(ctx, t) },
@@ -463,11 +448,10 @@ func TestAccRDSProxy_tags(t *testing.T) {
 		CheckDestroy:             testAccCheckProxyDestroy(ctx),
 		Steps: []resource.TestStep{
 			{
-				Config: testAccProxyConfig_tags1(rName, acctest.CtKey1, acctest.CtValue1),
+				Config: testAccProxyConfig_name(rName, rName),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckProxyExists(ctx, resourceName, &dbProxy),
-					resource.TestCheckResourceAttr(resourceName, acctest.CtTagsPercent, acctest.Ct1),
-					resource.TestCheckResourceAttr(resourceName, acctest.CtTagsKey1, acctest.CtValue1),
+					resource.TestCheckResourceAttr(resourceName, "tags.#", "0"),
 				),
 			},
 			{
@@ -476,20 +460,10 @@ func TestAccRDSProxy_tags(t *testing.T) {
 				ImportStateVerify: true,
 			},
 			{
-				Config: testAccProxyConfig_tags2(rName, acctest.CtKey1, acctest.CtValue1Updated, acctest.CtKey2, acctest.CtValue2),
-				Check: resource.ComposeTestCheckFunc(
-					testAccCheckProxyExists(ctx, resourceName, &dbProxy),
-					resource.TestCheckResourceAttr(resourceName, acctest.CtTagsPercent, acctest.Ct2),
-					resource.TestCheckResourceAttr(resourceName, acctest.CtTagsKey1, acctest.CtValue1Updated),
-					resource.TestCheckResourceAttr(resourceName, acctest.CtTagsKey2, acctest.CtValue2),
-				),
-			},
-			{
-				Config: testAccProxyConfig_tags1(rName, acctest.CtKey2, acctest.CtValue2),
+				Config: testAccProxyConfig_tags(rName, key, value),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckProxyExists(ctx, resourceName, &dbProxy),
-					resource.TestCheckResourceAttr(resourceName, acctest.CtTagsPercent, acctest.Ct1),
-					resource.TestCheckResourceAttr(resourceName, acctest.CtTagsKey2, acctest.CtValue2),
+					resource.TestCheckResourceAttr(resourceName, "tags.foo", value),
 				),
 			},
 		},
@@ -585,7 +559,7 @@ func testAccCheckProxyExists(ctx context.Context, n string, v *types.DBProxy) re
 	}
 }
 
-func testAccProxyConfig_base(rName string) string {
+func testAccProxyBaseConfig(rName string) string {
 	return fmt.Sprintf(`
 # Secrets Manager setup
 
@@ -675,14 +649,14 @@ resource "aws_subnet" "test" {
 }
 
 func testAccProxyConfig_basic(rName string) string {
-	return acctest.ConfigCompose(testAccProxyConfig_base(rName), fmt.Sprintf(`
+	return testAccProxyBaseConfig(rName) + fmt.Sprintf(`
 resource "aws_db_proxy" "test" {
   depends_on = [
     aws_secretsmanager_secret_version.test,
     aws_iam_role_policy.test
   ]
 
-  name                   = %[1]q
+  name                   = "%[1]s"
   debug_logging          = false
   engine_family          = "MYSQL"
   idle_client_timeout    = 1800
@@ -697,19 +671,23 @@ resource "aws_db_proxy" "test" {
     iam_auth    = "DISABLED"
     secret_arn  = aws_secretsmanager_secret.test.arn
   }
+
+  tags = {
+    Name = "%[1]s"
+  }
 }
-`, rName))
+`, rName)
 }
 
 func testAccProxyConfig_name(rName, nName string) string {
-	return acctest.ConfigCompose(testAccProxyConfig_base(rName), fmt.Sprintf(`
+	return testAccProxyBaseConfig(rName) + fmt.Sprintf(`
 resource "aws_db_proxy" "test" {
   depends_on = [
     aws_secretsmanager_secret_version.test,
     aws_iam_role_policy.test
   ]
 
-  name                   = %[1]q
+  name                   = "%[2]s"
   engine_family          = "MYSQL"
   role_arn               = aws_iam_role.test.arn
   vpc_security_group_ids = [aws_security_group.test.id]
@@ -722,18 +700,18 @@ resource "aws_db_proxy" "test" {
     secret_arn  = aws_secretsmanager_secret.test.arn
   }
 }
-`, nName))
+`, rName, nName)
 }
 
 func testAccProxyConfig_debugLogging(rName string, debugLogging bool) string {
-	return acctest.ConfigCompose(testAccProxyConfig_base(rName), fmt.Sprintf(`
+	return testAccProxyBaseConfig(rName) + fmt.Sprintf(`
 resource "aws_db_proxy" "test" {
   depends_on = [
     aws_secretsmanager_secret_version.test,
     aws_iam_role_policy.test
   ]
 
-  name           = %[1]q
+  name           = "%[1]s"
   debug_logging  = %[2]t
   engine_family  = "MYSQL"
   role_arn       = aws_iam_role.test.arn
@@ -746,18 +724,18 @@ resource "aws_db_proxy" "test" {
     secret_arn  = aws_secretsmanager_secret.test.arn
   }
 }
-`, rName, debugLogging))
+`, rName, debugLogging)
 }
 
 func testAccProxyConfig_idleClientTimeout(rName string, idleClientTimeout int) string {
-	return acctest.ConfigCompose(testAccProxyConfig_base(rName), fmt.Sprintf(`
+	return testAccProxyBaseConfig(rName) + fmt.Sprintf(`
 resource "aws_db_proxy" "test" {
   depends_on = [
     aws_secretsmanager_secret_version.test,
     aws_iam_role_policy.test
   ]
 
-  name                = %[1]q
+  name                = "%[1]s"
   idle_client_timeout = %[2]d
   engine_family       = "MYSQL"
   role_arn            = aws_iam_role.test.arn
@@ -770,18 +748,18 @@ resource "aws_db_proxy" "test" {
     secret_arn  = aws_secretsmanager_secret.test.arn
   }
 }
-`, rName, idleClientTimeout))
+`, rName, idleClientTimeout)
 }
 
 func testAccProxyConfig_requireTLS(rName string, requireTls bool) string {
-	return acctest.ConfigCompose(testAccProxyConfig_base(rName), fmt.Sprintf(`
+	return testAccProxyBaseConfig(rName) + fmt.Sprintf(`
 resource "aws_db_proxy" "test" {
   depends_on = [
     aws_secretsmanager_secret_version.test,
     aws_iam_role_policy.test
   ]
 
-  name           = %[1]q
+  name           = "%[1]s"
   require_tls    = %[2]t
   engine_family  = "MYSQL"
   role_arn       = aws_iam_role.test.arn
@@ -794,18 +772,18 @@ resource "aws_db_proxy" "test" {
     secret_arn  = aws_secretsmanager_secret.test.arn
   }
 }
-`, rName, requireTls))
+`, rName, requireTls)
 }
 
 func testAccProxyConfig_roleARN(rName, nName string) string {
-	return acctest.ConfigCompose(testAccProxyConfig_base(rName), fmt.Sprintf(`
+	return testAccProxyBaseConfig(rName) + fmt.Sprintf(`
 resource "aws_db_proxy" "test" {
   depends_on = [
     aws_secretsmanager_secret_version.test,
     aws_iam_role_policy.test2
   ]
 
-  name           = %[1]q
+  name           = "%[1]s"
   engine_family  = "MYSQL"
   role_arn       = aws_iam_role.test2.arn
   vpc_subnet_ids = aws_subnet.test[*].id
@@ -818,8 +796,10 @@ resource "aws_db_proxy" "test" {
   }
 }
 
+# IAM setup
+
 resource "aws_iam_role" "test2" {
-  name               = %[2]q
+  name               = "%[2]s"
   assume_role_policy = data.aws_iam_policy_document.assume.json
 }
 
@@ -827,18 +807,18 @@ resource "aws_iam_role_policy" "test2" {
   role   = aws_iam_role.test.id
   policy = data.aws_iam_policy_document.test.json
 }
-`, rName, nName))
+`, rName, nName)
 }
 
 func testAccProxyConfig_vpcSecurityGroupIDs(rName, nName string) string {
-	return acctest.ConfigCompose(testAccProxyConfig_base(rName), fmt.Sprintf(`
+	return testAccProxyBaseConfig(rName) + fmt.Sprintf(`
 resource "aws_db_proxy" "test" {
   depends_on = [
     aws_secretsmanager_secret_version.test,
     aws_iam_role_policy.test
   ]
 
-  name                   = %[1]q
+  name                   = "%[1]s"
   engine_family          = "MYSQL"
   role_arn               = aws_iam_role.test.arn
   vpc_security_group_ids = [aws_security_group.test2.id]
@@ -853,25 +833,21 @@ resource "aws_db_proxy" "test" {
 }
 
 resource "aws_security_group" "test2" {
-  name   = %[2]q
+  name   = "%[2]s"
   vpc_id = aws_vpc.test.id
-
-  tags = {
-    Name = %[2]q
-  }
 }
-`, rName, nName))
+`, rName, nName)
 }
 
 func testAccProxyConfig_authDescription(rName, description string) string {
-	return acctest.ConfigCompose(testAccProxyConfig_base(rName), fmt.Sprintf(`
+	return testAccProxyBaseConfig(rName) + fmt.Sprintf(`
 resource "aws_db_proxy" "test" {
   depends_on = [
     aws_secretsmanager_secret_version.test,
     aws_iam_role_policy.test
   ]
 
-  name                   = %[1]q
+  name                   = "%[1]s"
   engine_family          = "MYSQL"
   role_arn               = aws_iam_role.test.arn
   vpc_security_group_ids = [aws_security_group.test.id]
@@ -879,23 +855,23 @@ resource "aws_db_proxy" "test" {
 
   auth {
     auth_scheme = "SECRETS"
-    description = %[2]q
+    description = "%[2]s"
     iam_auth    = "DISABLED"
     secret_arn  = aws_secretsmanager_secret.test.arn
   }
 }
-`, rName, description))
+`, rName, description)
 }
 
 func testAccProxyConfig_authIAMAuth(rName, iamAuth string) string {
-	return acctest.ConfigCompose(testAccProxyConfig_base(rName), fmt.Sprintf(`
+	return testAccProxyBaseConfig(rName) + fmt.Sprintf(`
 resource "aws_db_proxy" "test" {
   depends_on = [
     aws_secretsmanager_secret_version.test,
     aws_iam_role_policy.test
   ]
 
-  name                   = %[1]q
+  name                   = "%[1]s"
   engine_family          = "MYSQL"
   role_arn               = aws_iam_role.test.arn
   require_tls            = true
@@ -905,22 +881,22 @@ resource "aws_db_proxy" "test" {
   auth {
     auth_scheme = "SECRETS"
     description = "test"
-    iam_auth    = %[2]q
+    iam_auth    = "%[2]s"
     secret_arn  = aws_secretsmanager_secret.test.arn
   }
 }
-`, rName, iamAuth))
+`, rName, iamAuth)
 }
 
 func testAccProxyConfig_authSecretARN(rName, nName string) string {
-	return acctest.ConfigCompose(testAccProxyConfig_base(rName), fmt.Sprintf(`
+	return testAccProxyBaseConfig(rName) + fmt.Sprintf(`
 resource "aws_db_proxy" "test" {
   depends_on = [
     aws_secretsmanager_secret_version.test,
     aws_iam_role_policy.test
   ]
 
-  name                   = %[1]q
+  name                   = "%[1]s"
   engine_family          = "MYSQL"
   role_arn               = aws_iam_role.test.arn
   vpc_security_group_ids = [aws_security_group.test.id]
@@ -944,7 +920,7 @@ resource "aws_db_proxy" "test" {
 }
 
 resource "aws_secretsmanager_secret" "test2" {
-  name                    = %[2]q
+  name                    = "%[2]s"
   recovery_window_in_days = 0
 }
 
@@ -952,46 +928,18 @@ resource "aws_secretsmanager_secret_version" "test2" {
   secret_id     = aws_secretsmanager_secret.test2.id
   secret_string = "{\"username\":\"db_user\",\"password\":\"db_user_password\"}"
 }
-`, rName, nName))
-}
-
-func testAccProxyConfig_tags1(rName, tagKey1, tagValue1 string) string {
-	return testAccProxyConfig_base(rName) + fmt.Sprintf(`
-resource "aws_db_proxy" "test" {
-  depends_on = [
-    aws_secretsmanager_secret_version.test,
-    aws_iam_role_policy.test
-  ]
-
-  name                   = %[1]q
-  engine_family          = "MYSQL"
-  role_arn               = aws_iam_role.test.arn
-  vpc_security_group_ids = [aws_security_group.test.id]
-  vpc_subnet_ids         = aws_subnet.test[*].id
-
-  auth {
-    auth_scheme = "SECRETS"
-    description = "test"
-    iam_auth    = "DISABLED"
-    secret_arn  = aws_secretsmanager_secret.test.arn
-  }
-
-  tags = {
-    %[2]q = %[3]q
-  }
-}
-`, rName, tagKey1, tagValue1)
+`, rName, nName)
 }
 
-func testAccProxyConfig_tags2(rName, tagKey1, tagValue1, tagKey2, tagValue2 string) string {
-	return testAccProxyConfig_base(rName) + fmt.Sprintf(`
+func testAccProxyConfig_tags(rName, key, value string) string {
+	return testAccProxyBaseConfig(rName) + fmt.Sprintf(`
 resource "aws_db_proxy" "test" {
   depends_on = [
     aws_secretsmanager_secret_version.test,
     aws_iam_role_policy.test
   ]
 
-  name                   = %[1]q
+  name                   = "%[1]s"
   engine_family          = "MYSQL"
   role_arn               = aws_iam_role.test.arn
   vpc_security_group_ids = [aws_security_group.test.id]
@@ -1005,9 +953,8 @@ resource "aws_db_proxy" "test" {
   }
 
   tags = {
-    %[2]q = %[3]q
-    %[4]q = %[5]q
+    %[2]s = "%[3]s"
   }
 }
-`, rName, tagKey1, tagValue1, tagKey2, tagValue2)
+`, rName, key, value)
 }
